File: with_elixir/lib/day2.ex
defmodule Day2 do
  def part1(content) do
    from_report_to_int(content)
    |> Enum.count(fn x -> is_valid?(x) end)
  end

  def part2(content) do
    from_report_to_int(content)
    |> Enum.count(fn x -> is_valid?(x) || is_valid_allowing_one_exception?(x) end)
  end

  defp from_report_to_int(content) do
    content
    |> Enum.map(fn x -> String.split(x) |> Enum.map(&String.to_integer/1) end)
  end

  defp is_valid?(report) do
    is_sorted?(report) and is_gap_ok?(report)
  end

  defp is_sorted?(report) do
    report == Enum.sort(report) || report == Enum.sort(report, :desc)
  end

  defp is_gap_ok?(report) do
    report
    |> Enum.zip(tl(report))
    |> Enum.all?(fn {a, b} -> abs(a - b) >= 1 and abs(a - b) <= 3 end)
  end

  defp is_valid_allowing_one_exception?(report) do
    report
    |> remove_one()
    |> Enum.any?(fn xs -> is_sorted?(xs) and is_gap_ok?(xs) end)
  end

  defp remove_one([]), do: []

  defp remove_one([x | xs]) do
    [xs | Enum.map(remove_one(xs), &[x | &1])]
  end
end

File: with_haskell/lib/Day2.hs
module Day2 (part1, part2) where
import Data.List (sort, zip, tail, sortBy)
import qualified Data.Ord
import Data.Ord (comparing)

part1 :: [String] -> Int
part1 = foldl checkValidity 0
  where
    checkValidity acc = (+) acc . fromEnum . isValid . map read . words
    isValid xs = isSorted xs && isGapOk xs

part2 :: [String] -> Int
part2 = foldl checkValidity 0
  where
    checkValidity acc = (+) acc . fromEnum . isValid . map read . words
    isValid xs = (isSorted xs && isGapOk xs) || isValidAllowingOneException xs

    isValidAllowingOneException :: [Int] -> Bool
    isValidAllowingOneException xs = any (\x -> isSorted x && isGapOk x) (removeOneAt xs)


isSorted :: [Int] -> Bool
isSorted xs = xs == sort xs || xs == sortBy (comparing Data.Ord.Down) xs

isGapOk :: [Int] -> Bool
isGapOk xs = all (\(x, y) -> abs (x - y) >= 1 && abs (x - y) <= 3) (zip xs (tail xs))

removeOneAt :: [a] -> [[a]]
removeOneAt [] = []
removeOneAt (x:xs) = xs : map (x:) (removeOneAt xs)